drivers/sensor/bosch/bmi08x/bmi08x_accel_decoder.c

```diff
@@ -0,0 +1,226 @@
+};
+
+struct frame_len {
+	enum bmi08x_accel_fifo_header header;
+	uint8_t len;
+} fifo_frame_len[] = {
+	{.header = BMI08X_ACCEL_FIFO_FRAME_ACCEL, .len = 7},
+	{.header = BMI08X_ACCEL_FIFO_FRAME_SKIP, .len = 2},
+	{.header = BMI08X_ACCEL_FIFO_FRAME_TIME, .len = 4},
+	{.header = BMI08X_ACCEL_FIFO_FRAME_CONFIG, .len = 2},
+	{.header = BMI08X_ACCEL_FIFO_FRAME_DROP, .len = 2},
+	{.header = BMI08X_ACCEL_FIFO_FRAME_EMPTY, .len = 2},
+};
+
+void bmi08x_accel_encode_header(const struct device *dev, struct bmi08x_accel_encoded_data *edata,
+			       bool is_streaming, uint16_t buf_len)
+{
+	struct bmi08x_accel_data *data = dev->data;
+	uint64_t cycles;
+
+	if (sensor_clock_get_cycles(&cycles) == 0) {
+		edata->header.timestamp = sensor_clock_cycles_to_ns(cycles);
+	} else {
+		edata->header.timestamp = 0;
+	}
+	edata->header.has_accel = true;
+	edata->header.range = data->range;
+	edata->header.chip_id = data->accel_chip_id;
+	edata->header.is_streaming = is_streaming;
+	edata->header.sample_count = is_streaming ? data->stream.fifo_wm : 1;
+	edata->header.buf_len = buf_len;
+}
+
+static int bmi08x_decoder_get_frame_count(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
+					  uint16_t *frame_count)
+{
+	const struct bmi08x_accel_encoded_data *edata =
+		(const struct bmi08x_accel_encoded_data *)buffer;
+
+	if (!edata->header.has_accel || chan_spec.chan_idx != 0) {
+		return -ENODATA;
+	}
+
+	if (chan_spec.chan_type != SENSOR_CHAN_ACCEL_XYZ) {
+		return -EINVAL;
+	}
+
+	*frame_count = edata->header.sample_count;
+	return 0;
+}
+
+static int bmi08x_decoder_get_size_info(struct sensor_chan_spec chan_spec, size_t *base_size,
+					size_t *frame_size)
+{
+	if (chan_spec.chan_idx != 0 || chan_spec.chan_type != SENSOR_CHAN_ACCEL_XYZ) {
+		return -EINVAL;
+	}
+
+	*base_size = sizeof(struct sensor_three_axis_data);
+	*frame_size = sizeof(struct sensor_three_axis_sample_data);
+	return 0;
+}
+
+static inline void fixed_point_from_encoded_data(const uint16_t encoded_payload[3], uint8_t shift,
+						 uint32_t fsr_value_g, q31_t output[3])
+{
+	for (size_t i = 0 ; i < 3 ; i++) {
+		int64_t raw_value;
+
+		raw_value = sign_extend_64(encoded_payload[i], 15);
+		raw_value = (raw_value * fsr_value_g << (31 - 5 - 15)) * SENSOR_G / 1000000;
+
+		output[i] = raw_value;
+	}
+}
+
+static inline int bmi08x_decode_one_shot(const struct bmi08x_accel_encoded_data *edata,
+					 uint32_t *fit, struct sensor_three_axis_data *data_output)
+{
+	uint32_t fsr_value_g = (edata->header.chip_id == BMI085_ACCEL_CHIP_ID) ? 2 : 3;
+
+	if (*fit != 0) {
+		return -ENODATA;
+	}
+
+	/** Bits we need to represent the integer part of FSR in m/s2:
+	 * - 2 - 3 G (19.6 - 29.4 m/s2) = 5 bits.
+	 * - 4 - 6 G (39.2 - 58.8 m/s2) = 6 bits.
+	 * - 8 - 12 G (78.4 - 117.6 m/s2) = 7 bits.
+	 * - 16 - 24 G (156.8 235.2 m/s2) = 8 bits.
+	 */
+	data_output->shift = 5 + edata->header.range;
+	data_output->header.reading_count = 1;
+	data_output->header.base_timestamp_ns = edata->header.timestamp;
```

Comment:
@ubieda coming late to this review, but you sure about this here? The timestamp is recorded when the gpio event happens, which would be the last sample that comes out of the FIFO, but this is recording a time of the very last sample it will read and samples are somehow in the 'future', Shouldn't this be calculating the time of the 'first' sample similiar to how it is done with the lsm6dsv16x seen here: https://github.com/zephyrproject-rtos/zephyr/blob/0f5e03f1fcba4326baf4507c343f3609bf32c524/drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_decoder.c#L306